### PR TITLE
placement: push rules check down to rule manager

### PR DESCRIPTION
**Pull request by xhebox**
### What problem does this PR solve?

It is redundant to check rules two times: one in `api/rule.go`, another in `rule_manager.go`. This PR moves all checks in `api/rule.go` into `adjustRule` from `rule_manager.go`, which will recude one `O(N)` loop for all setting rules methods.

And now we can maintain only one single unique rule validation method `adjustRule` than two copies. My next PR on #3299 depends this PR.

### Check List

Tests

- Unit test
- Integration test

### Release note

- No release note

### CODE REVIEW

**Review by rleungx — APPROVED** [review]
LGTM

File: pkg/errs/errno.go [hunk 1]
```diff
@@ -228,11 +228,6 @@ var (
 	ErrNetstatTCPSocks = errors.Normalize("TCP socks error", errors.RFCCodeText("PD:netstat:ErrNetstatTCPSocks"))
 )
 
-// hex error
-var (
```
Comment (rleungx) [inline pkg/errs/errno.go:232]: I noticed why remove this error?

Comment (xhebox) [reply pkg/errs/errno.go:232]: It is weird to wrap the hex error again with `ErrRuleContent`, looks like `[placement:PD:xx]invalid rule, [placement:PD:xx] invalid hex` or so... And it is only used here, in this function, after merging two check functions.

Comment (rleungx) [reply pkg/errs/errno.go:232]: But actually, we need to use an individual error type for those error come from the third-party package.

Comment (xhebox) [reply pkg/errs/errno.go:232]: Then, what about not using this error in adjustRule? And I will add it back.

Comment (xhebox) [reply pkg/errs/errno.go]: What about now?

Comment (rleungx) [reply pkg/errs/errno.go]: IMO, we can keep it as before and return the BadRequest when the error is ErrHexDecoding

Comment (xhebox) [reply pkg/errs/errno.go]: One more error check..? OK, later I will send a fix.

Comment (xhebox) [reply pkg/errs/errno.go]: Done

File: server/api/rule.go [hunk 2]
```diff
@@ -80,14 +78,13 @@ func (h *ruleHandler) SetAll(w http.ResponseWriter, r *http.Request) {
 	if err := apiutil.ReadJSONRespondError(h.rd, w, r.Body, &rules); err != nil {
 		return
 	}
-	for _, rule := range rules {
-		if err := h.checkRule(rule); err != nil {
+	if err := cluster.GetRuleManager().SetKeyType(h.svr.GetConfig().PDServerCfg.KeyType).
+		SetRules(rules); err != nil {
+		if errs.ErrRuleContent.Equal(err) && errs.ErrHexDecodingString.Equal(err) {
```
Comment (rleungx) [inline server/api/rule.go:83]: `||` ?

Comment (xhebox) [reply server/api/rule.go]: Yeah. sorry. I am doing some other things.

**Review by Yisaer — APPROVED** [review]
LGTM

**Pull request merged**